src/components/verification-grid/verification-grid.ts

```diff
@@ -292,6 +338,194 @@ export class VerificationGridComponent extends AbstractComponent(LitElement) {
+
+    return { width, height };
+  }
+
+  private willFitTileSize(containerSize: Size, gridShape: GridShape): boolean {
+    const minimumTileWidth: Pixel = this.minimumGridSize.width;
+    const minimumTileHeight: Pixel = this.minimumGridSize.height;
+
+    const proposedTileSize = this.tileSize(containerSize, gridShape);
+
+    const meetsMinimumWidth = proposedTileSize.width >= minimumTileWidth;
+    const meetsMinimumHeight = proposedTileSize.height >= minimumTileHeight;
+
+    return meetsMinimumHeight && meetsMinimumWidth;
+  }
+
+  private gridAspectRatioSimilarity(containerSize: Size, gridShape: GridShape): number {
+    const targetAspectRatio = containerSize.width / containerSize.height;
+    const candidateSizeAspectRatio = gridShape.columns / gridShape.rows;
+
+    const distance = Math.abs(targetAspectRatio - candidateSizeAspectRatio);
+    const maxRatio = Math.max(targetAspectRatio, candidateSizeAspectRatio);
+    const similarity = 1 - distance / maxRatio;
+
+    return similarity;
+  }
+
+  private computeGridShape(target: number): GridShape {
+    if (target === 1) {
+      return { columns: 1, rows: 1 };
+    } else if (target === 4) {
+      // TODO: Find out why 2x2 is not being preferred in our current algorithm
+      // and fix it so that we don't have to use this hack
+      //
+      // our current implementation works well for most grid sizes
+      // however, for a grid target of 4, it will prefer a 4x1 grid over a 2x2
+      // this is because 2x2 does not meet our aspect ratio threshold
+      // this is expected because we don't usually want a 1:1 aspect ratio
+      // for a 16:9 screen
+      // however, in the case of a target of 4, we want a 2x2 grid
+      // I've used this hack to force a 2x2 grid for a target of 4
+      // when the grid containers aspect ratio is close to 16:9
+      const screenSize = window.screen;
+      const aspectRatio = screenSize.width / screenSize.height;
+
+      // we allow for a bit of deviation from the exact 16:9 aspect ratio so
+      // similar aspect ratios also get the nicer grid size for a target of 4
+      const matchingAspectRatio = 16 / 9;
+      const matchingDeviation = 0.2;
+
+      if (Math.abs(aspectRatio - matchingAspectRatio) <= matchingDeviation) {
+        return { columns: 2, rows: 2 };
+      }
+    }
+
+    const containerSize = this.gridContainer.getBoundingClientRect();
+    let refinedTarget = target;
+    const testedTargets: number[] = [];
+
+    // some numbers e.g. 9 have one factor such as 3x3 which creates a aspect
+    // ratio of 1.00
+    // using a 3x3 grid for a 16:9 screen is not ideal because it will create
+    // a lot of dead space
+    // therefore, we have a threshold that we have to meet. If we do not meet
+    // the threshold, we keep increasing the target until we find a grid shape
+    // that meets the similarity percentage threshold
+    const targetThreshold = 0.5;
```

Comment:
I suspect the threshold is playing havoc here. What happens if you remove this?

Reply:
Have you tried selecting factors with the most similar aspect ratio?

Reply:
By optimizing for tile area, we are indirectly optimizing for aspect ratio.

---

The optimal grid size is determined by two factors:

- Conditions that **must** be met
- and conditions that we **should** to optimize

We need pre-conditions so that grid sizes such as 3x3 are completely rejected and the we optimize to 5x2